rust/lance-linalg/src/simd/f16.c

```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __ARM_NEON
+#define LANES 4
+#define UNROLL_COUNT 2
+#endif  // __ARM_NEON
+
+#ifdef __x86_64__
+#define LANES 64
+#define UNROLL_COUNT 4
+#endif  // X86_64
+
+/// Works on NEON + FP16 or AVX512FP16
+//
+// Please make sure run "cargo bench --bench norm_l2" on both Apple Silicon and
+// X86_64, before you change this function.
+
+#ifdef __ARM_NEON
+
+// The NEON version is 40% faster than the x86_64 version on Apple M2 Max.
```

Comment:
Is there anything NEON specific about this?  This version is 40% faster on my x86_64 as well.  The (slower) version GCC comes up with is something like...

```
  _Float16 sum = 0;
  _Float16* itr = data + (dimension / 8 * 8) - 8;
  if (dimension % 8 == 0) goto A;
  if (dimension % 8 == 1) goto B;
  if (dimension % 8 == 2) goto C;
  if (dimension % 8 == 3) goto D;
  if (dimension % 8 == 4) goto E;
  if (dimension % 8 == 5) goto F;
  if (dimension % 8 == 6) goto G;
  if (dimension % 8 == 7) goto H;
 H:
  sum += itr[14] * itr[14];
 G:
  sum += itr[13] * itr[13];
 F:
  sum += itr[12] * itr[12];
 E:
  sum += itr[11] * itr[11];
 D:
  sum += itr[10] * itr[10];
 C:
  sum += itr[9] * itr[9];
 B:
  sum += itr[8] * itr[8];
 A:
  sum += itr[7] * itr[7];
  sum += itr[6] * itr[6];
  sum += itr[5] * itr[5];
  sum += itr[4] * itr[4];
  sum += itr[5] * itr[5];
  sum += itr[4] * itr[4];
  sum += itr[3] * itr[3];
  sum += itr[2] * itr[2];
  sum += itr[1] * itr[1];
  sum += itr[0] * itr[0];
  itr -= 8;
  if (itr == data) return sum;
  goto A;
```

Even without recognizing the clang pragmas, gcc will take your version and generate faster code.

Reply:
It explicitly say that there is no data dependency on `sum`. IIUC, ARM has more registers than x86_64, but shorter CPU pipelines.  Also avx512 needs 32 `f16`s to kick in. 

What is your OS / CPU @westonpace